src/storage/src/sink/kafka.rs

```diff
@@ -336,16 +365,40 @@ impl TransactionalProducer {
             timestamp: None,
             delivery_opaque: (),
         };
-        let key_size = key.map(|k| k.len()).unwrap_or(0);
-        let value_size = value.map(|k| k.len()).unwrap_or(0);
-        let record_size = u64::cast_from(key_size + value_size);
-        self.statistics.inc_messages_staged_by(1);
-        self.staged_messages += 1;
-        self.statistics.inc_bytes_staged_by(record_size);
-        self.staged_bytes += record_size;
+
         self.producer.send(record).map_err(|(e, _)| e)
     }
 
+    /// Updates the reported statistics for this sink based on the statistics received from
+    /// librdkafka. Since `librdkafka` does not differentiate between committed and uncomitted
+    /// updates the function accepts a flag to indicate if it should assume that the messages
+    /// reported as produced are also committed.
+    fn update_stats(&mut self, assume_committed: bool) {
+        let stats = self.stats_receiver.borrow_and_update();
+        let messages_staged = u64::try_from(stats.txmsgs).unwrap();
+        let bytes_staged = u64::try_from(stats.txmsg_bytes).unwrap();
+
+        self.statistics.set_messages_buffered(stats.msg_cnt);
+        self.statistics
+            .inc_messages_staged_by(messages_staged - self.prev_messages_staged);
```

Comment:
Turning these underlying counters into gauges could save you the math here! (I'd understand if that's more of a refactor than you want to bite off, though.)

Reply:
Actually that's what I originally set out to do but read that counters have some additional benefits, namely that when a process restarts and they reset to zero prometheus understands that and if you're asking for a rate it does the right thing at the discontinuity point. I don't know if this is important or not but that's why I went through the trouble of computing diffs